starter-api/src/main/java/io/micronaut/starter/api/create/github/GitHubCreateController.java

```diff
@@ -0,0 +1,80 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.starter.api.create.github;
+
+import edu.umd.cs.findbugs.annotations.NonNull;
+import edu.umd.cs.findbugs.annotations.Nullable;
+import io.micronaut.http.HttpHeaders;
+import io.micronaut.http.MediaType;
+import io.micronaut.http.annotation.Controller;
+import io.micronaut.http.annotation.Get;
+import io.micronaut.http.annotation.Header;
+import io.micronaut.scheduling.TaskExecutors;
+import io.micronaut.scheduling.annotation.ExecuteOn;
+import io.micronaut.starter.api.TestFramework;
+import io.micronaut.starter.api.create.github.client.v3.GitHubRepository;
+import io.micronaut.starter.application.ApplicationType;
+import io.micronaut.starter.options.BuildTool;
+import io.micronaut.starter.options.JdkVersion;
+import io.micronaut.starter.options.Language;
+
+import javax.validation.constraints.Pattern;
+import java.util.List;
+
+/**
+ * GitHub create controller.
+ *
+ * @author Pavol Gressa
+ * @since 2.1
+ */
+@Controller
+@ExecuteOn(TaskExecutors.IO)
+public class GitHubCreateController implements GitHubCreateOperation {
+
+    private final GitHubCreateService gitHubCreateService;
+
+    public GitHubCreateController(GitHubCreateService gitHubCreateService) {
+        this.gitHubCreateService = gitHubCreateService;
+    }
+
+    /**
+     * Creates an application in GitHub repository, generating a json file as the response.
+     *
+     * @param type     The application type The application type
+     * @param name     The name of the application The name of the application
+     * @param features The features The chosen features
+     * @param build    The build type (optional, defaults to Gradle)
+     * @param test     The test framework (optional, defaults to JUnit)
+     * @param lang     The language (optional, defaults to Java)
+     * @return A json containing the generated application details.
+     */
+    @Override
+    @Get(uri = "/github/{type}/{name}{?features,lang,build,test,javaVersion,code,state}", produces = MediaType.APPLICATION_JSON)
+    public GitHubCreateDTO createApp(
+            @NonNull ApplicationType type,
+            @Pattern(regexp = "[\\w\\d-_\\.]+") String name,
+            @Nullable List<String> features,
+            @Nullable BuildTool build,
+            @Nullable TestFramework test,
+            @Nullable Language lang,
+            @Nullable JdkVersion javaVersion,
+            @NonNull String code,
+            @NonNull String state,
+            @Nullable @Header(HttpHeaders.USER_AGENT) String userAgent) {
+        GitHubRepository repository = gitHubCreateService.creatApp(type, name, features, build, test, lang, javaVersion, code, state, userAgent);
+        return new GitHubCreateDTO(repository.getUrl(), repository.getCloneUrl(), repository.getHtmlUrl());
```

Comment:
@pgressa this is the one other place where we may need to return a redirect. I guess it all depends on how we set up the GH OAuth App.  

If the the OAuth app is being set to do the callback to the **starter** here this will need to be a redirect, instead of JSON.  However if the GH OAuth App is getting configured to do the redirect to to **launch** the I will be able to just parse the redirect from GH on the UI and then make a call to this ending and poll accordingly.

I'm fine either way, just want to make sure we're on the same page.

Reply:
@eahrold sure, check this commit https://github.com/micronaut-projects/micronaut-starter/pull/495/commits/1754c5164d2301e412f5082fab449bd638ec030d

Reply:
I'm not sure this is working as expected.  I've got an Oauth App Set up in Github

<img width="822" alt="Screen Shot 2020-10-06 at 10 07 58 PM" src="https://user-images.githubusercontent.com/3740088/95283091-86e40c00-0820-11eb-804b-c45d1b83356f.png">

Then from a browser I go to http://localhost:8080/github/default/demo?build=gradle&lang=java

at this point I get the correct redirect to GH https://github.com/login/oauth/authorize?scope=user%2Crepo&redirect_uri=http%3A%2F%2Flocalhost%3A8080%2Fgithub%2Fdefault%2Fdemo&state=d47764fb-f1d1-454a-8ef7-f0023a0dd7cc&client_id=bfd389a24260f5fed94e. (side note I think this is missing the other url parameters from the initial request such as **build** and **lang** above)

The app authorizes and the

n calls back to the `redirect_url` which successfully generates the Repo on GitHub (awesome) but the it responds to the browser with a json string. 

I think the final response [Here](https://github.com/micronaut-projects/micronaut-starter/blob/1754c5164d2301e412f5082fab449bd638ec030d/starter-api/src/main/java/io/micronaut/starter/api/create/github/GitHubCreateController.java#L100) also needs to be a redirect back to the `launch`.  

I've seen it where the original request origin is bound to the unique`state` for the final redirect lookup at this point. I'm not sure the best approach.

Here's a screen cast of how it's working now.  (I'd already authorized the app so it just skips over the GH auth that part of the flow.)

![Redirect mov](https://user-images.githubusercontent.com/3740088/95284514-f7405c80-0823-11eb-9d84-655462745436.gif)

Reply:
True, since these are independent apps, so I have to redirect the user back to launch UI. What is your preference of the redirect target so you can render the success/failure of the action?

> side note I think this is missing the other url parameters from the initial request such as build and lang above

Will write test for it just for sure.

Reply:
It could be simply to the root with a query parameter of the repo

Http://flow.origin?url=...&cloneUrl=...&htmlUrl=...

Reply:
ok will do

Reply:
@eahrold ok so now if the conditions are met, which given the circumstances are everytime at this moment, the endpoint basically returns just 307's . First is for authorisation callback and second is to launcher. Check https://github.com/micronaut-projects/micronaut-starter/pull/495/commits/f6783dbee56562c7d03800b47461b6bace8ed6ca

Reply:
Happy path is working great.  But still getting JSON when the error is thrown due to duplicate repo ( which definitely happens)

https://github.com/micronaut-projects/micronaut-starter/pull/495/files#diff-c59a5ffd3f9a9658e42605a866b52763R140

We may want to throw a custom GitHubException so we can Narrowly handle and redirect with something like ‘?error=someerrormessage’

Reply:
sure will do that

Reply:
@eahrold error handling should be there.

Reply:
@pgressa Looks good.  Here's the ui companion.
https://github.com/micronaut-projects/micronaut-starter-ui/pull/14